interfaces/mount/change.go

```diff
@@ -0,0 +1,121 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package mount
+
+import (
+	"path"
+	"sort"
+	"strings"
+)
+
+// Action represents a mount action (mount, remount, unmount, etc).
+type Action string
+
+const (
+	Mount   Action = "mount"
+	Unmount Action = "umount"
+	// Remount when needed
+)
+
+// Change describes a change to the mount table (action and the entry to act on).
+type Change struct {
+	Entry  Entry
+	Action Action
+}
+
+// NeededChanges computes the changes required to change current to desired mount entries.
+//
+// The current and desired profiles is a fstab like list of mount entries. The
+// lists are processed and a "diff" of mount changes is produced. The mount
+// changes, when applied in order, transform the current profile into the
+// desired profile.
+func NeededChanges(currentProfile, desiredProfile []Entry) []Change {
+	// Copy both as we will want to mutate them.
+	current := make([]Entry, len(currentProfile))
+	copy(current, currentProfile)
+	desired := make([]Entry, len(desiredProfile))
+	copy(desired, desiredProfile)
+
+	// Clean the directory part of both profiles. This is done so that we can
+	// easily test if a given directory is a subdirectory with
+	// strings.HasPrefix coupled with an extra slash character.
+	for i := range current {
+		current[i].Dir = path.Clean(current[i].Dir)
+	}
+	for i := range desired {
+		desired[i].Dir = path.Clean(desired[i].Dir)
+	}
+
+	// Sort both lists by directory name with implicit trailing slash.
+	sort.Sort(byMagicDir(current))
+	sort.Sort(byMagicDir(desired))
+
+	// Construct a desired directory map.
+	// Maps from a directory to a pointer to an Entry from the desired list.
+	desiredMap := make(map[string]*Entry)
+	for i := range desired {
+		desiredMap[desired[i].Dir] = &desired[i]
+	}
+
+	// Reuse map, indexed by Entry.Dir.
+	// All reused entries will not be unmounted or mounted. Reused entries must
+	// naturally exist in the desired and current maps or no reuse is possible.
+	reuse := make(map[string]bool)
+
+	// See if there are any directories that we can reuse. Go over all the
+	// entries in the current entries and if there's an identical desired entry
+	// then mark this directory / entry for reuse.
+	//
+	// Don't reuse any children if their parent changes.
+	var skipPrefix string
+	for i := range current {
+		dir := current[i].Dir
+		if skipPrefix != "" && strings.HasPrefix(dir, skipPrefix) && dir[len(skipPrefix)] == '/' {
```

Comment:
`if skipDir != "" && strings.HasPrefix(dir, skipDir) {`

Let's guarantee insideDir has a slash instead, otherwise the logic is hoping the call site agrees with its conventions, with no actual verification.